### PR TITLE
Changes while reviewing your work on the VVwE PR

### PR DESCRIPTION
**Pull request by davidrusu**
Making changes to the code as I reviewed your work from the VVwE PR: https://github.com/rust-crdt/rust-crdt/pull/75

Putting this up for visibility, will ping you when it's ready for you to look at

### CODE REVIEW

**Comment by davidrusu** [comment]
@xldenis Can you take a look at this change, if you think it's good, merge it in to your branch, then rebase your branch on `rust-crdt:master`?

I had made some improvements to the `Dot` interface that I want to make use of here.

Then we can continue getting this work ready to merge

File: src/vvwe.rs [hunk 1]
```diff
@@ -4,73 +4,77 @@
...
 
-use crate::vclock::{Actor, Dot};
+    fn no_exceptions(&self, clk: LogTime) -> bool {
+        !self.exceptions.contains(&clk)
+    }
+}
 
 pub trait CausalOp<A> {
-
+    /// TODO: result should be a VClock<A> since an op could be dependant on a few different msgs
```
Comment (xldenis) [inline src/vvwe.rs:77]: oh thats an interesting generalization!

File: src/vvwe.rs [hunk 2]
```diff
@@ -80,65 +84,69 @@ pub trait CausalOp<A> {
...
                 }
             }
             None => {
                 // Ok so we're not causally constrained, but maybe we already saw an associated
                 // causal operation? If so let's just delete the pair
-                match self.buffer.remove(&msg.dot()) {
-                    Some(_) => None,
-                    None => Some(msg),
+                match self.buffer.remove(&op.dot()) {
+                    Some(_) => None, // we are dropping the dependent op! that can't be right
```
Comment (xldenis) [inline src/vvwe.rs:123]: thats correct but only for a situation where messages are inserts and updates. 

If we had a `Delete(IX)` in the buffer and we see the matching `Insert(IX, C)` then they cancel out.

Comment (xldenis) [reply src/vvwe.rs:123]: I think we could generalize this by having some sort of multiplication for operations. (Do CRDT operations necessarily form a semigroup?)

**Comment by xldenis** [comment]
I left some comments explaining why things are but these changes look good.

File: src/vvwe.rs (hunk 2 above)
Comment (davidrusu) [reply src/vvwe.rs:123]: I don't think cancelling out is a concern of the causality barrier. My intuition is just to keep the barrier as a sort of buffer to re-order and release operations as they are ready to be consumed by the underlying causal algorithm.

Comment (davidrusu) [reply src/vvwe.rs:123]: Consider that the underlying algorithm already  has the logic in place to "cancel" ops if that's how they should behave. 

ie. the LSEQ understands how to "cancel" out an "insert" with a "delete", this logic is in the `apply` method.

To also have this logic in the CausalityBarrier would be duplicating logic and potentially causing bugs if the cancellation logic is not in sync with the underlying algorithm.

Comment (xldenis) [reply src/vvwe.rs:123]: In the current implementation where causal pairs are exactly insertions and deletions i think it's fine / correct but if we want to add support for batch operations / more complex relations then it'll have to move to vec<op> for sure anyways and cancelling out inside the VVwE is probably not a great idea anymore.

**Pull request merged**